Add sale_allotment

clonedagain opened the pull request:
This aims to implement partner allotment as discussed in https://github.com/OCA/sale-workflow/issues/206 and https://github.com/odoo/odoo/issues/5371

Forked from the initial implementation by @Openies here: https://www.odoo.com/apps/modules/8.0/line_ship/
- [x] check license for OCA compatibility (OK: AGPL)
- [x] check whether @openies agrees to OCA's CLA

Improvements made (or to make):
- [x] rename the module to `sale_allotment`
- [x] access `procurement_id.sale_line_id.address_allotment_id` directly instead of hijacking `procurement.partner_dest_id`
- [x] adapt to OCA's conventions (README, PEP8...)
- [x] add context to attach quick-created partners as contacts of the customer, with "use the company's address" checked by default
- [x] add the allotment partner to the sale line's treeview (those who don't use other advanced features can't see the field because it's only in the form view)
- [x] make sure all the pickings share the same parameters

Commit:
Added Lineship Module

Commit:
Changed Category

Commit:
Changed summary

Openies committed:
Changed index.html

Commit:
Rename line_ship to sale_allotment and improve it for OCA adoption

Keep the same parameters for all the pickings
Directly use the sale line's allotment. The procurement's partner is used for drop shipping, we should probably not hijack it.
Add the allotment partner to the sale line's treeview. Those who don't use other advanced features can't see the field otherwise because it's only in the form view.
Create allotment partners as children of the customer, with type='other' by default. Creating them as type 'delivery' by default would be error-prone, since they would always be proposed for the next orders.
Conform to OCA conventions
Add tests for sale_allotment

NB: We don't group allotments on the same picking if the partner uses the parent adress. It was suggested but actually a bit naive given the very loose coupling we now have between stock moves and pack operations. It turned out to be mostly unworkable, because stock workers don't have the dispatch information at the right place anymore.

clonedagain commented:
I think this is ready for review, pending @Openies approval of the CLA.

StefanRijnhart commented:
Nice module! I confirm @Openis or Anil Makwana cannot be found in the OCA database yet.

Openies commented:
@Stefan Thanks for the comment.
yes, we have never submitted our module to OCA.

On Wed, Sep 9, 2015 at 12:00 AM, Stefan Rijnhart (Opener) <
notifications@github.com> wrote:

> Nice module! I confirm @Openis or Anil Makwana cannot be found in the OCA
> database yet.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/207#issuecomment-138662048.

## 

Openies
Ingenious Effective Solutions

clonedagain commented:
@Openies have you had time yet to decide if you agree with OCA's CLA ? it's especially important now that v9 is released.
It'd be perfectly fine by me whether you agree or not: depending on your choice I'd just make a PR to your own branch instead of OCA's.

Openies commented:
Hello Lionel,

Yes, We are working on that, will revert you soon with CLA.

On Tue, Oct 6, 2015 at 1:22 PM, Lionel Sausin notifications@github.com
wrote:

> @Openies https://github.com/Openies have you had time yet to decide if
> you agree with OCA's CLA ? it's especially important now that v9 is
> released.
> It'd be perfectly fine by me whether you agree or not: depending on your
> choice I'd just make a PR to your own branch instead of OCA's.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/207#issuecomment-145774519.

## 

Openies
Ingenious Effective Solutions

Openies commented:
Hello Lionel,

We have submitted our Signed OCA CLA.

Openies http://Openies.com.

On Fri, Oct 9, 2015 at 2:18 PM, Openies India openies.com@gmail.com wrote:

> Hello Lionel,
> 
> Yes, We are working on that, will revert you soon with CLA.
> 
> On Tue, Oct 6, 2015 at 1:22 PM, Lionel Sausin notifications@github.com
> wrote:
> 
> > @Openies https://github.com/Openies have you had time yet to decide if
> > you agree with OCA's CLA ? it's especially important now that v9 is
> > released.
> > It'd be perfectly fine by me whether you agree or not: depending on your
> > choice I'd just make a PR to your own branch instead of OCA's.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/OCA/sale-workflow/pull/207#issuecomment-145774519.
> 
> ## 
> 
> Openies
> Ingenious Effective Solutions

## 

Openies
Ingenious Effective Solutions

clonedagain commented:
I think I fixed the small coding convention nitpicking, I hope Travis goes green .

Openies commented:
@clonedagain We have pushed Lineship module to https://github.com/Openies/OdooLineShip/tree/9.0
please have a look.
Thanks!

clonedagain commented:
@Openies sorry we have no plan to move to v9. Maybe you can propose to migrate/enhance this module directly at OCA's if it gets merged.

clonedagain commented:
@lbellier can you review this please?

lbellier commented:
+1

StefanRijnhart commented on sale_allotment/models/stock_move.py:
```diff
@@ -0,0 +1,67 @@
+        pick_obj = self.pool.get('stock.picking')
+        for key in partner_dict.keys():
+            move = move_rec[0]
+            if partner_dict.keys().index(key) == 0:
```
What is this line supposed to do? It looks like you rely on the order of the keys of the dictionary which is unsafe when using regular dictionaries. You need collections.OrderedDict for that.

clonedagain replied:
The standard provides us with a single picking already created.
The way I understand it, it's just a way to assign this picking to one of the groups, and then create new pickings for the other groups.
So the order is not important.

StefanRijnhart commented on sale_allotment/models/stock_move.py:
```diff
@@ -0,0 +1,67 @@
+            if not partner_dict.get(move.partner_id.id, False):
+                partner_dict.update({move.partner_id.id: [move.id]})
+            else:
+                partner_dict.get(move.partner_id.id).append(move.id)
```
You could replace these four lines with

```
partner_dict.setdefault(move.partner_id.id, []).append(move.id)
```

StefanRijnhart commented on sale_allotment/models/stock_move.py:
```diff
@@ -0,0 +1,67 @@
+            else:
+                partner_dict.get(move.partner_id.id).append(move.id)
+        pick_obj = self.pool.get('stock.picking')
+        for key in partner_dict.keys():
```
Would be clearer if you renamed 'key' to 'partner_id'

StefanRijnhart commented on sale_allotment/models/stock_move.py:
```diff
@@ -0,0 +1,67 @@
+                }
+                pick = pick_obj.create(cr, uid, values, context=context)
+                self.write(
+                    cr, uid, partner_dict.get(key), {'picking_id': pick},
```
No need to use 'get()'

lbellier commented:
In fact ! This module create pickings without the same header.
It search for user / default parameters only for the first picking, and without for the following others...

lbellier committed:
[IMP]: move column address_allotment_id from "before name" to "after product_id".

clonedagain commented:
@lbellier @StefanRijnhart I think I addressed your remarks: I've rewritten a part of this module to duplicate the pickings instead of creating new ones and losing the parameters.

clonedagain commented:
I fixed an api v7/v8 compatibility problem in the code, and added a test to make coverage progress.
Can we merge?

lbellier commented:
:+1: :+1: !

StefanRijnhart commented:
Thanks for the update! One last question ;-) Why is there a domain on partners of type 'other'? That would exclude delivery addresses (but contacts might be fine too?).

clonedagain commented:
@StefanRijnhart, initially we had set the default type and the domain to 'delivery' but users found it error-prone, so we changed both to 'other' instead.
But you made a point, so I've relaxed the domain to "other, delivery or contact" and kept 'contact' as the default.
@lbellier do you agree?

StefanRijnhart commented:
Thanks. Would you consider squashing the commits per author? It is a bit noisy at the moment in the commits section.

lbellier commented:
@clonedagain, Yes, i agree.

clonedagain commented:
@StefanRijnhart: Rebased and squashed, but I've respected @Openies' commits in case they want to merge with their branch.

StefanRijnhart commented:
Thanks! You did include some pyc files however.

clonedagain commented:
Oh, the magic of git rebase. I think it's fixed now.

StefanRijnhart commented:
Thanks! :+1:

clonedagain commented:
Can this be merged please?
